cordapp-upgrades/rpc-client/src/main/kotlin/net/obligation/rpcClient/RpcClient.kt

```diff
@@ -0,0 +1,101 @@
+package net.obligation.rpcClient
+
+import net.corda.client.rpc.CordaRPCClient
+import net.corda.client.rpc.CordaRPCConnection
+import net.corda.client.rpc.RPCException
+import net.corda.core.node.services.vault.AttachmentQueryCriteria
+import net.corda.core.utilities.NetworkHostAndPort
+import net.corda.core.utilities.contextLogger
+import net.corda.core.utilities.seconds
+import org.apache.activemq.artemis.api.core.ActiveMQConnectionTimedOutException
+import org.apache.activemq.artemis.api.core.ActiveMQSecurityException
+import org.apache.activemq.artemis.api.core.ActiveMQUnBlockedException
+
+class RpcClient(config: RpcClientConfig) {
+
+    companion object {
+        val log = contextLogger()
+    }
+
+    private val partyToConnection = config.partyToRpcPort.map { (party, networkHostAndPort) ->
+        val nodeAddress = NetworkHostAndPort.parse(networkHostAndPort)
+        val connection = establishConnectionWithRetry(nodeAddress)
+        Pair(party, connection)
+    }.toMap().toMutableMap()
+
+    private val connections = partyToConnection.values
+
+    fun getConnection(party: String): CordaRPCConnection {
+        return partyToConnection.getValue(party)
+    }
+
+    fun uploadAttachmentsToAllNodes() {
+        // The attachments cannot simply be opened here, as the first node the attachment is uploaded to will close it.
+        // Instead, create a mapping of attachment ids to a connection that can be used to open it.
+        val attachmentLocations = connections.flatMap { connection ->
+            val ids = connection.proxy.queryAttachments(AttachmentQueryCriteria.AttachmentsQueryCriteria(), null)
+            ids.map { Pair(it, connection) }
+        }.toMap()
+
+        connections.forEach { connection ->
+            attachmentLocations.forEach { (id, conn) ->
+                if (!connection.proxy.attachmentExists(id)) {
+                    val attachment = conn.proxy.openAttachment(id)
+                    connection.proxy.uploadAttachment(attachment)
+                }
+            }
+        }
+    }
+
+    fun closeAllConnections() {
+        connections.forEach { it.close() }
```

Comment:
Each close call must be surrounded by a try-catch to ensure every connection is closed if there's an exception thrown by an earlier connection in the list.